### ISSUE TITLE
scenario_seqCurve.py 没跑起来

### ISSUE THREAD

**Issue by fmscole** [AskyJx/xDeepLearning#1]
scenario_mnist.py可以正常跑起来，
但是 scenario_seqCurve.py 我没跑起来，不知道您是不是没有提交最新代码？

**Comment by AskyJx** [comment]
你好，升级了优化方法，scenario没有同步更新。
现在已经同步了。

**Comment by fmscole** [comment]
@AskyJx  谢谢！

**Comment by fmscole** [comment]
有没有计划写attention层？

**Comment by AskyJx** [comment]
在看paper

**Comment by fmscole** [comment]
有个问题想跟你探讨一下，关于bn层。
前段时间我正好在学习bn，但看你的代码，我感觉跟我理解的不太一致：
1）63行：mu = np.mean(x, axis=0)
      我的理解是：如果input是 B*C*W*H 的话，mu = np.mean(x, axis=(0,2,3)) 
2）91行：dx = 1./N* ivar * (N*dxhat-np.sum(dxhat, axis=1) -xhat*np.sum(dxhat*xhat,axis=1))
     似乎应该是：dx = 1./N* ivar * (N*dxhat-np.sum(dxhat, axis=1) - ivar*xhat*np.sum(dxhat*xhat,axis=1))
    xhat 前少乘了一个ivar。
我学得比较粗糙，不知道你怎么看？
另外，能否写一个test，与pytorch或者TensorFlow的正式算法做一个对比？
下面是我试着写的一个测试（不过要注释掉bn.py中好多代码），代码极其粗糙简陋，见谅：
```
"""
Created: May 2018
@author: JerryX
Find more : https://www.zhihu.com/people/xu-jerry-82
"""

import numpy as np
import time

import logging.config

# create logger
import os
exec_abs = os.getcwd()
log_conf = exec_abs + '/config/logging.conf'
logging.config.fileConfig(log_conf)
logger = logging.getLogger('bn')


# batch normalization layer
# with optimizer adopter
class BNLayer(object):

    def __init__(self,LName,eps,miniBatchesSize,channel,i_size,activator,optimizerCls,optmParams, dataType):
        self.name = LName
        self.miniBatchesSize = miniBatchesSize
        self.i_size = i_size
        self.channel = channel
        self.activator = activator
        self.eps = eps
        # self.optimizerObj = optimizerCls(optmParams, dataType)
        self.dataType = dataType
        self.beta = np.zeros((channel,i_size,i_size),dataType)
        self.gamma = np.ones((channel,i_size,i_size),dataType)
        self.mu_accum = np.zeros((channel,i_size,i_size),dataType)
        self.var_accum = np.ones((channel,i_size,i_size),dataType)
        self.decay_accum = 0.95  # 滑动平均衰减率
        self.cache = ()
        self.out = []
        self.deltaPrev = []  
        self.deltaOri = [] 

    # 滑动平均衰减累计mini-batch的mean和variance
    def mov_avg_accum(self,mu,var):
        decay = self.decay_accum

        self.mu_accum = decay * self.mu_accum + (1-decay) * mu
        self.var_accum = decay * self.var_accum + (1-decay) * var

    def inference(self,x):
        oriOut = self.bnForward_inf(x,self.gamma,self.beta,self.eps)
        self.out =oriOut # self.activator.activate(oriOut)
        return self.out

    def fp(self,x):
        oriOut,self.cache = self.bnForward_tr(x,self.gamma,self.beta,self.eps)
        self.out =oriOut # self.activator.activate(oriOut)
        return self.out

    # for training
    def bnForward_tr(self, x, gamma, beta, eps):

      mu = np.mean(x, axis=1)
      xmu = x - mu
      var = np.mean(xmu **2, axis = 1)

      self.mov_avg_accum(mu, var)

      ivar = 1./np.sqrt(var + eps)
      xhat = xmu * ivar
      out = gamma*xhat + beta

      cache = (xhat,gamma,ivar)

      return out, cache

    # 已优化，不使用分步骤方式
    def bnBackward(self,dout, cache):

      st = time.time()
      logger.debug('bnBk start')

      xhat,gamma,ivar = cache

      N = dout.shape[0]

      dbeta = np.sum(dout, axis=1)
      dgamma = np.sum(dout*xhat, axis=1)

      dxhat = dout * gamma
      dx = 1./N* ivar * (N*dxhat-np.sum(dxhat, axis=1) - xhat*np.sum(dxhat*xhat,axis=1))

      logger.debug('bnBk end, %s s', (time.time()-st) )

      return dx, dgamma, dbeta


    # for predict, Unbiased Estimation
    def bnForward_inf(self, x, gamma, beta, eps):

      N = self.miniBatchesSize

      var = N / (N - 1) * self.var_accum
      tx = (x - self.mu_accum) /np.sqrt(var + eps)
      out = gamma * tx + beta

      return out

    def bp(self,input,delta):
        self.deltaOri =delta # self.activator.bp(delta, self.out)

        self.deltaPrev, dgamma, dbeta = self.bnBackward(self.deltaOri, self.cache)

        weight=(self.gamma,self.beta)
        dweight = (dgamma,dbeta)
        # self.optimizerObj.getUpdWeights(weight,dweight, lrt)

        return self.deltaPrev

if __name__=="__main__":
  import torch
  import torch.nn as nn
  import numpy as np
  from torch.autograd import Variable as V

  bn1=BNLayer("LName",1e-5,3,2,2,None,None,None, np.float32)
  m = nn.BatchNorm2d(2,momentum=None,  affine=False)
  x=np.array(np.random.randint(-10,10,24)).reshape((3,2,2,2))
  input=V(torch.Tensor(x),requires_grad = True)
  print(input[0,0])
  output = m(input,)
  output2 = bn1.inference(x)
  print("torch---",output[0,0])
  print("my------",output2[0,0])
  # print(output.requires_grad)
  dy=list([[[[ 1.3028,  0.5017],
            [-0.8432, -0.2807]],

          [[-0.4656,  0.2773],
            [-0.7269,  0.1338]]],


          [[[-3.1020, -0.7206],
            [ 0.4891,  0.2446]],

          [[ 0.2814,  2.2664],
            [ 0.8446, -1.1267]]],


          [[[-2.4999,  1.0087],
            [ 0.6242,  0.4253]],

          [[ 2.5916,  0.0530],
            [ 0.5305, -2.0655]]]])
  # dnb=bn1.bp(x,np.array(dy))
  print("")
  print("")
  # print("my---",dnb[0,0])
  dy=torch.tensor(dy)

  output.backward(dy)
  print("torch--",input.grad[0,0])
```
观察下来，前向结果不一致，反向传播没能测试成功，注释掉了。
不知道是我的代码有问题，还是算法本身有问题？

**Comment by AskyJx** [comment]
1）63行：mu = np.mean(x, axis=0)
你的理解是对的，paper上建议卷积层在BWH维度做BN，即axis=(0,2,3)。BN的源码也增加了对卷积层的支持。感谢你指出这个错误。
2）91：dx = 1./N ivar * (Ndxhat-np.sum(dxhat, axis=1) -xhatnp.sum(dxhatxhat,axis=1))

**Comment by AskyJx** [comment]
1）63行：mu = np.mean(x, axis=0)
你的理解是对的，paper上建议卷积层在BWH维度做BN，即axis=(0,2,3)，所以原来BN层对卷积层的归一化处理和paper是不一致的，现已更新。感谢你指出这个问题。
2）91：dx = 1./N ivar * (Ndxhat-np.sum(dxhat, axis=1) -xhatnp.sum(dxhatxhat,axis=1))
 右边原来有个ivar^2，拆成两个ivar，一个作为公因式提到前面，剩下一个和(x_i-\mu)搭配成标准化变量\hat x_i了。

**Comment by fmscole** [comment]
谢谢



---原始邮件---
发件人: "Jerry X"<notifications@github.com>
发送时间: 2019年8月7日(星期三) 晚上6:39
收件人: "AskyJx/xDeepLearning"<xDeepLearning@noreply.github.com>;
抄送: "fmscole"<1047170189@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [AskyJx/xDeepLearning] scenario_seqCurve.py 没跑起来 (#1)



1）63行：mu = np.mean(x, axis=0)
 你的理解是对的，paper上建议卷积层在BWH维度做BN，即axis=(0,2,3)，所以原来BN层对卷积层的归一化处理和paper是不一致的，现已更新。感谢你指出这个问题。
 2）91：dx = 1./N ivar * (Ndxhat-np.sum(dxhat, axis=1) -xhatnp.sum(dxhatxhat,axis=1))
 右边原来有个ivar^2，拆成两个ivar，一个作为公因式提到前面，剩下一个和(x_i-\mu)搭配成标准化变量\hat x_i了。

**Comment by fmscole** [comment]
请问对ctc有关注吗，想请教几个问题



---原始邮件---
发件人: "Jerry X"<notifications@github.com>
发送时间: 2019年8月7日(星期三) 晚上6:39
收件人: "AskyJx/xDeepLearning"<xDeepLearning@noreply.github.com>;
抄送: "fmscole"<1047170189@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [AskyJx/xDeepLearning] scenario_seqCurve.py 没跑起来 (#1)



1）63行：mu = np.mean(x, axis=0)
 你的理解是对的，paper上建议卷积层在BWH维度做BN，即axis=(0,2,3)，所以原来BN层对卷积层的归一化处理和paper是不一致的，现已更新。感谢你指出这个问题。
 2）91：dx = 1./N ivar * (Ndxhat-np.sum(dxhat, axis=1) -xhatnp.sum(dxhatxhat,axis=1))
 右边原来有个ivar^2，拆成两个ivar，一个作为公因式提到前面，剩下一个和(x_i-\mu)搭配成标准化变量\hat x_i了。
